### PR TITLE
Reproducible

### PR DESCRIPTION
**Pull request by james-w**
Make the image build reproducible, so if you run it twice with the same
inputs you will get the same tarball out.

This isn't used for anything useful yet.

### CODE REVIEW

**Comment by chipironcin** [comment]
Check some comments and reasons at https://github.com/bitnami/minideb/pull/15#issuecomment-277052115

**Comment by james-w** [comment]
Updated to produce the same image ids by cheating a bit on the image creation date.

The other downside of this approach that I have found is that dockerhub uses push
time rather than creation time to report when an image last changed, even if the push
didn't change the image. We might want to add back in a guard around pushing that
checks whether the image is the same and skips the push.

**Comment by chipironcin** [comment]
So even if we manage to reproduce an image build, if we push it it will be shown in dockerHub as updated so we need #15 to avoid pushing if not changed. Right?

BTW this PR is getting complicated, including the cheat 😛 
I guess it is ok but we may want to write down (in a section in the README for example) why and how are we doing these kind of things. If I looked at the files for the first time my head would explode 😂

**Comment by james-w** [comment]
My understanding is that if we push the same thing again the web UI will say "updated a few seconds ago", but pulls will report that nothing has changed and not do anything, and `docker images` will show the correct build date for the image.

I think that's fine, but if you care about the web UI showing the right thing, or if it turns out to be confusing then we can add something back that doesn't try and push if the image is the same.

About the documentation, I plan to write an engineering blogpost explaining it. At the least we can
then link to that.

Which particular areas do you think warrant some more explanation in comments?

File: buildall [hunk 1]
```diff
@@ -13,19 +14,61 @@ GCR_BASENAME=gcr.io/bitnami-containers/minideb
 
 mkdir -p build
 
+import() {
```
Comment (chipironcin) [inline buildall]: Why do we need this wrapping function?

File: buildall [hunk 2]
```diff
@@ -13,19 +14,61 @@ GCR_BASENAME=gcr.io/bitnami-containers/minideb
...
+log() {
+    echo "$@" >&2
+}
+
 for DIST in $DISTS; do
    [ -f debootstrap/$DIST ] || (echo "buildall: Unknown distribution: $DIST" && exit 1)
-   echo "============================================"
-   echo "Building $BASENAME:$DIST"
-   echo "============================================"
+   current_ts="$(date -u +%Y-%m-%dT%H:%M:%S.%NZ)"
```
Comment (chipironcin) [inline buildall]: I would add a comment here detailing the steps:
```
# We want to compare built images and pulled image between them
# Docker will assign the same id to images that it considers are the same
# Docker consider two images are the same if metadata like creation date and checksum are the same
# 1. Pull image from Dockerhub and save creation date and image_id
# 2. Build image locally and import it setting creation date to the pulled one
# 3. Build again image locally and import it setting creation date to the pulled one
# 4. Compare locally built images id. Error if they are not the same (Docker thinks images are different, thanks to checksum)
# 5. Compare built image id with pulled image id. Both will have same creation date but may differ in checksum so ids may be different

File: buildall [hunk 3]
```diff
@@ -13,19 +14,61 @@ GCR_BASENAME=gcr.io/bitnami-containers/minideb
...
+   if docker pull $BASENAME:$DIST > /dev/null; then
+       target_ts="$(docker inspect $BASENAME:$DIST | jq --raw-output ".[0].Created")"
```
Comment (chipironcin) [inline buildall]: naming suggestion: pulled_image_ts

File: buildall [hunk 4]
```diff
@@ -13,19 +14,61 @@ GCR_BASENAME=gcr.io/bitnami-containers/minideb
...
+   fi
+   log "============================================"
+   log "Building $BASENAME:$DIST"
+   log "============================================"
    ./mkimage build/$DIST.tar $DIST
-   IMPORTED=$(docker import --change "CMD /bin/bash" build/$DIST.tar)
-   echo "============================================"
-   echo "Running tests for $BASENAME:$DIST"
-   echo "============================================"
+   IMPORTED=$(import build/$DIST.tar "$target_ts")
```
Comment (chipironcin) [inline buildall]: Naming suggestion: built_image_id (following convention from pulled_image_id)

File: buildall [hunk 5]
```diff
@@ -13,19 +14,61 @@ GCR_BASENAME=gcr.io/bitnami-containers/minideb
...
+   log "============================================"
+   log "Running tests for $BASENAME:$DIST"
+   log "============================================"
    ./test $IMPORTED
+   log "============================================"
+   log "Rebuilding $BASENAME:$DIST to test reproducibility"
+   log "============================================"
+   ./mkimage build/${DIST}-repro.tar $DIST
+   REPRO=$(import build/${DIST}-repro.tar "$target_ts")
```
Comment (chipironcin) [inline buildall]: Naming suggestion: repro_image_id

File: buildall (hunk 4 above)
Comment (chipironcin) [inline buildall]: Naming suggestion: getImageIDWithCustomTimestamp or something that speaks by itself, I was super confused about the work `import` passing a tarball and a timestamp. Although using comments can help to understand the whole point behind `import` so it is not important to change it

File: dockerdiff [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
```
Comment (chipironcin) [inline dockerdiff]: Add little explanation about what does this do

File: import [hunk 7]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
```
Comment (chipironcin) [inline import]: Add little explanation about what does this do

File: import [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+set -u
+set -o pipefail
+
+CONF_TEMPLATE='{"architecture":"amd64","comment":"from Bitnami with love","config":{"Hostname":"","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":null,"Cmd":["/bin/bash"],"Image":"","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"container_config":{"Hostname":"","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":null,"Cmd":null,"Image":"","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"created":"%TIMESTAMP%","docker_version":"1.13.0","history":[{"created":"%TIMESTAMP%","comment":"from Bitnami with love"}],"os":"linux","rootfs":{"type":"layers","diff_ids":["sha256:%LAYERSUM%"]}}'
+MANIFEST_TEMPLATE='[{"Config":"%CONF_SHA%.json","RepoTags":null,"Layers":["%LAYERSUM%/layer.tar"]}]'
+
+SOURCE=${1:?Specify the tarball to import}
+TIMESTAMP=${2:?Specify the timestamp to use}
+
+import() {
```
Comment (chipironcin) [inline import]: why wrapping this in a function if this script has only one purpose?

File: buildall (hunk 1 above)
Comment (james-w) [inline buildall]: Left over, I will remove it.

File: buildall (hunk 3 above)
Comment (james-w) [inline buildall]: It's not the pulled image timestamp if we didn't pull, I chose "target" as it is the one we are aiming for.

File: import (hunk 8 above)
Comment (james-w) [inline import]: To allow using `local` in case the script grows more in future.

**Pull request merged**